Implement a well-behaved stacktrace printer.

bbockelm opened the pull request:
This implements a stack trace printer which is safe to call
from a signal handler (ROOT's stack trace printer is not async-signal
safe; see 'man 7 signal' for more information about async-signal
safety).

Further, on Linux, this utilizes the clone library call directly;
clone has the advantage of avoiding the pthread_atfork handlers.
We have observed that jemalloc currently registers AS-unsafe atfork
handlers, which would cause a deadlock if we use fork().

The commit addresses real deadlocks observed in the Tier-0.

bbockelm committed:
Implement a well-behaved stacktrace printer.

This implements a stack trace printer which is safe to call
from a signal handler (ROOT's stack trace printer is not async-signal
safe; see 'man 7 signal' for more information about async-signal
safety).

Further, on Linux, this utilizes the clone library call directly;
clone has the advantage of avoiding the pthread_atfork handlers.
We have observed that jemalloc currently registers AS-unsafe atfork
handlers, which would cause a deadlock if we use fork().

cmsbuild commented:
A new Pull Request was created by @bbockelm (Brian Bockelman) for CMSSW_7_6_X.

Implement a well-behaved stacktrace printer.

It involves the following packages:

FWCore/Services


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

bbockelm commented:
@davidlange6 - I think this should fix the deadlocks observed at the Tier-0.

Basically, the patch is just very careful about avoiding memory allocations and most standard library functions from the context of the signal handler.  Note also I use the CLONE_VM to avoid memory allocation issues in the child process before execv is invoked.

FWIW - I did check the glibc implementation of clone() to validate it is AS-safe.

Dr15Jones commented:
please test

Dr15Jones commented:
Does this work on a Mac?

cmsbuild commented:
The tests are being triggered in jenkins.

sextonkennedy commented:
I asked google the MAC question and it came up with this:
http://stackoverflow.com/questions/24523118/pstack-a-process-on-osx-10-9
which sounds promising...

cmsbuild commented:
-1 
 
Tested at: cff10d347ca92d4408d5f4b5838180daf495a888
I found errors in the following unit tests: 
 
 ---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10084/8552/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/25435a576dc97780ea10d4848d358fe560af146d
https://github.com/cms-sw/cmssw/commit/7a8ff1cfcdffbcee072de9f5c49e685b6bad0bb2
https://github.com/cms-sw/cmssw/commit/ef332418be99d934dad32f3b7d841f446fcab40e
https://github.com/cms-sw/cmssw/commit/6fab1b82d68b8bfa1c969c957ebd929389b39f12
https://github.com/cms-sw/cmssw/commit/0156bbbaa8d91d5d84d544eb0232fbc0aaa4d7d5
https://github.com/cms-sw/cmssw/commit/6c999d9e4b66ac2afaf1f8d4b8157e28c0503cf2
https://github.com/cms-sw/cmssw/commit/d8f205c3e3eea13043e1404ffbe506cf13cee4e7
https://github.com/cms-sw/cmssw/commit/0c925f9681c4a3dfa25dde12054d9b7ab9ac047f
https://github.com/cms-sw/cmssw/commit/11106440c2505e179f53443d526a888da540c5da
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10084/8552/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10084/8552/git-merge-result

bbockelm committed:
Replace use of pstack with direct invocation of gdb.

This switches from the system pstack script to using gdb directly.
Ideally, this will be more usable on non-Linux platforms.

NOTE that this will pull the gdb from the environment (i.e., probably
CMSSW's gdb instead of the system one).  CMSSW's gdb can parse the
debug symbols we ship, meaning that the backtrace will include
niceties like filenames, line numbers, and decode arguments.

The downside of this is that CMSSW's gdb requires 10 minutes
and about 1GB RAM to generate the traceback.

bbockelm commented:
@Dr15Jones - It seems the Jenkins failures are unrelated.

@sextonkennedy - I switched from using 'pstack' to using the 'gdb' shipped by CMSSW; this should be available on all our platforms.  The upside of our `gdb` is we ship symbols, so tracebacks have an immense amount of debugging information in them.  The downside of our `gdb` is it appears to take >1GB RAM and 10 minutes to generate a traceback.

cmsbuild commented:
Pull request #10084 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

bbockelm commented:
@sextonkennedy - @smuzaffar clarified my comment above via HN.  The very slow GDB will only occur on releases where debug symbols are shipped.  I was doing my testing based on an IB; "normal" resources shouldn't have this issue.

Dr15Jones commented on FWCore/Services/src/InitRootHandlers.cc:
```diff
@@ -337,5 +414,17 @@ namespace edm {
+    void
+    InitRootHandlers::cachePidInfo()
+    {
+      assert(snprintf(pidString_, pidStringLength_-1, "gdb -quiet -p %d 2>&1 <<EOF |\n"
```
Don't do real work inside an assert since a compile switch can make the assert go away.

cmsbuild commented:
+1
Tested at: b6d0bae4c1c009fe51dafd75bbb148d1bff14cb5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10084/10967/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10084/10967/summary.html

bbockelm committed:
Toss exception instead of doing work in an assert.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

bbockelm commented:
@davidlange6 @Dr15Jones - what (if any) branches would you like this backported to?

bbockelm commented:
@davidlange6 - bump on the question above.  Would you like a backport to any other branch?

Dr15Jones commented:
please test
